arangod/Cluster/ClusterMethods.cpp

```diff
@@ -183,17 +207,27 @@ void addTransactionHeaderForShard(transaction::Methods const& trx,
     return;  // no need
   }
 
-  auto const& it = shardMap.find(shard);
-  if (it != shardMap.end()) {
-    if (it->second.empty()) {
-      THROW_ARANGO_EXCEPTION(TRI_ERROR_CLUSTER_BACKEND_UNAVAILABLE);
-    }
-    ServerID const& leader = it->second[0];
+  // If we are in a reading transaction and are supposed to read from followers
+  // then we need to send transaction begin headers not only to leaders, but
+  // sometimes also to followers. The `TransactionState` knows this and so
+  // we must consult `whichReplica` instead blindly taking the leader.
+  // Note that this essentially only happens in `getDocumentOnCoordinator`.
+  if (trx.state()->options().allowDirtyReads) {
+    ServerID const& leader = trx.state()->whichReplica(shard);
```

Comment:
`leader` should be renamed to `server` here, because it can also contain a follower.